### PR TITLE
[BRO] fix Slagstone Refinery trigger

### LINKED ISSUE #9755: [BRO] Slagstone Refinery is bugged.
[[Slagstone Refinery]] is quite buggy in its first implementation:

- Does not trigger on other non-token artifacts, but does trigger a token artifact (I would have add that missing `!` myself as a PR if that was the only issue)
- On sacrificing a token artifact, there is immediatly a new trigger in the stack, which creates a new powerstone. Leading to an infinite trigger + powerstone creation.
- Passing the turn with the Refinery in play leads to a breaking game error:
```
Game exception occurred: mage.MageException: Iterated player priority after game exception too often, game ends! Last error:
 null
Server version: 1.4.52-V1-beta9 (build: 2022-11-12 06:32)
mage.game.GameImpl.playPriority(GameImpl.java:1591)
mage.game.turn.Step.priority(Step.java:61)
mage.game.turn.Phase.playStep(Phase.java:184)
mage.game.turn.Phase.play(Phase.java:89)
mage.game.turn.Turn.play(Turn.java:132)
mage.game.GameImpl.playTurn(GameImpl.java:1056)
mage.game.GameImpl.play(GameImpl.java:966)
mage.game.GameImpl.start(GameImpl.java:942)
mage.server.game.GameWorker.call(GameWorker.java:34)
java.util.concurrent.FutureTask.run(FutureTask.java:266)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by Susucre**
fixes #9755 by rewriting the checkTrigger method.
One thing I did miss in the issue post was there was no check for the 'you control' part of the trigger.

### CODE REVIEW

File: Mage.Sets/src/mage/cards/s/SlagstoneRefinery.java [hunk 1]
```diff
@@ -59,10 +60,25 @@ public boolean checkEventType(GameEvent event, Game game) {
     @Override
     public boolean checkTrigger(GameEvent event, Game game) {
         ZoneChangeEvent zEvent = (ZoneChangeEvent) event;
-        return zEvent.getFromZone().match(Zone.BATTLEFIELD)
-                && (zEvent.getToZone().match(Zone.GRAVEYARD) || zEvent.getToZone().match(Zone.EXILED))
-                && zEvent.getTargetId().equals(getSourceId())
-                || (zEvent.getTarget().isArtifact(game) && !(zEvent.getTarget() instanceof PermanentToken));
+
+        // Filters events for battlefield => graveyard or exile => graveyard.
+        if (zEvent.getFromZone() == Zone.BATTLEFIELD && (zEvent.getToZone().match(Zone.GRAVEYARD) || zEvent.getToZone().match(Zone.EXILED))) {
```
Comment (weirddan455) [inline Mage.Sets/src/mage/cards/s/SlagstoneRefinery.java:65]: Should be consistent (use either == or .match).  I think == is fine here.  match checks for Zone.ALL but I don't think that's relevant in this context.

File: Mage.Sets/src/mage/cards/s/SlagstoneRefinery.java [hunk 2]
```diff
@@ -59,10 +60,25 @@ public boolean checkEventType(GameEvent event, Game game) {
...
+            if(zEvent.getTargetId().equals(getSourceId())) return true; // {this}
+            else {                                                                                  // another
+                UUID targetId = zEvent.getTargetId();
+                Permanent permanent = game.getPermanent(targetId);
+                if (permanent == null) {
+                    permanent = (Permanent) game.getLastKnownInformation(targetId, Zone.BATTLEFIELD);
+                }
```
Comment (weirddan455) [inline Mage.Sets/src/mage/cards/s/SlagstoneRefinery.java:72]: This isn't needed.  Just use `zEvent.getTarget().getControllerId()`.

**Review by weirddan455 — CHANGES_REQUESTED** [review]

**Comment by Susucre** [comment]
@weirddan455 thanks for the review.
Code should be cleaner now.

**Review by weirddan455 — APPROVED** [review]
Thanks, looks good.

**Pull request merged**